std/complex.d

```diff
@@ -973,6 +973,179 @@ Complex!T tan(T)(Complex!T z) @safe pure nothrow @nogc
     assert(ceqrel(tan(complex(0.0, 1.0)), complex(0.0, std.math.tanh(1.0))) >= double.mant_dig - 2);
 }
 
+/**
+    Inverse trigonometric functions on complex numbers.
+
+    Params: z = A complex number.
+    Returns: The arcsine, arccosine and arctangent of `z`, respectively.
+*/
+Complex!T asin(T)(Complex!T z)  @safe pure nothrow @nogc
+{
+    auto ash = asinh(Complex!T(-z.im, z.re));
+    return Complex!T(ash.im, -ash.re);
+}
+
+///
+@safe pure nothrow unittest
+{
+    import std.math : isClose, PI;
+    assert(asin(complex(0.0)) == 0.0);
+    assert(isClose(asin(complex(0.5L, 0)), PI / 6));
+}
+
+/// ditto
+Complex!T acos(T)(Complex!T z)  @safe pure nothrow @nogc
+{
+    static import std.math;
+    auto as = asin(z);
+    return Complex!T(std.math.PI_2 - as.re, as.im);
+}
+
+///
+@safe pure nothrow unittest
+{
+    import std.math : isClose, PI, std_math_acos = acos;
+    assert(acos(complex(0.0)) == std_math_acos(0.0));
+    assert(isClose(acos(complex(0.5)), PI / 3));
+}
+
+/// ditto
+Complex!T atan(T)(Complex!T z) @safe pure nothrow @nogc
+{
+    static import std.math;
+    const T re2 = z.re * z.re;
+    const T x = 1.0 - re2 - z.im * z.im;
+
+    T num = z.im + 1.0;
+    T den = z.im - 1.0;
+
+    num = re2 + num * num;
+    den = re2 + den * den;
+
+    return Complex!T(0.5 * std.math.atan2(2.0 * z.re, x),
+                     0.25 * std.math.log(num / den));
```

Comment:
It's probably better to use `T(0.5)` etc. for the literals, to prevent temporary double-precision for `T == float`.

Reply:
Unless the compiler has learnt a new trick, I don't think that really makes a difference, as internally it's all `real` anyway.

Reply:
I don't know what you mean by 'internally' in this context. This doesn't even compile for `float`, due to ambiguous `atan2` call:
```
Error: std.math.atan2 called with argument types (double, const(float)) matches both:
/dlang/dmd/linux/bin64/../../src/phobos/std/math.d(1106):     std.math.atan2(real y, real x)
and:
/dlang/dmd/linux/bin64/../../src/phobos/std/math.d(1123):     std.math.atan2(float y, float x)
```

Reply:
Right, I didn't think of that.  I wonder how many other functions here might have issues due to lack of testing `Complex!float`.

Reply:
> Unless the compiler has learnt a new trick,

well `std.math.atan2` uses float now